File: real_benchmarks/data/real_c_rust_pairs/null_pointer.c
/*
 * NULL pointer divergence — inspired by CVE-2009-1897 (Linux kernel NULL deref)
 * and CVE-2018-1000001 (glibc realpath NULL deref).
 *
 * In C, dereferencing NULL is UB. Compilers can optimize away NULL checks
 * that follow a dereference (since UB means it "can't" be NULL).
 * In Rust, Option<&T> makes nullability explicit; raw pointer deref requires unsafe.
 */
#include <stddef.h>
#include <string.h>

struct Node {
    int value;
    struct Node *next;
};

/* CVE-style: dereference before NULL check (compiler may elide the check). */
int get_value_if_valid(struct Node *node) {
    int val = node->value;   /* dereference first — UB if node is NULL */
    if (node == NULL)         /* compiler may optimize this away */
        return -1;
    return val;
}

/* Linked list traversal: no NULL guard on inner access. */
int sum_list(struct Node *head) {
    int sum = 0;
    struct Node *cur = head;
    while (cur != NULL) {
        sum += cur->value;
        cur = cur->next;
    }
    return sum;
}

/* String length with potential NULL input (real pattern from glibc bugs). */
int string_length_or_default(const char *s) {
    if (s == NULL) return 0;
    return (int)strlen(s);
}

/* Double-pointer pattern: *pp may be NULL after lookup. */
int deref_result(struct Node **pp) {
    struct Node *p = *pp;
    return p->value;  /* UB if *pp was NULL */
}

File: real_benchmarks/data/real_c_rust_pairs/null_pointer.rs
/*
 * Rust translation of NULL pointer patterns.
 * Rust uses Option<&T> / Option<Box<T>> — no implicit NULL.
 * Raw pointer deref requires unsafe block.
 */

pub struct Node {
    pub value: i32,
    pub next: Option<Box<Node>>,
}

/// In Rust, node is Option — we must pattern match.
/// No UB: the compiler enforces the check.
pub fn get_value_if_valid(node: Option<&Node>) -> i32 {
    match node {
        Some(n) => n.value,
        None => -1,
    }
}

/// Linked list sum — Option makes the traversal safe.
pub fn sum_list(head: Option<&Node>) -> i32 {
    let mut sum = 0i32;
    let mut cur = head;
    while let Some(node) = cur {
        sum = sum.wrapping_add(node.value);
        cur = node.next.as_deref();
    }
    sum
}

/// String length: Option<&str> eliminates NULL ambiguity.
pub fn string_length_or_default(s: Option<&str>) -> i32 {
    match s {
        Some(string) => string.len() as i32,
        None => 0,
    }
}

/// Double-reference pattern: Rust uses Option to encode absence.
pub fn deref_result(pp: &Option<Box<Node>>) -> i32 {
    match pp {
        Some(p) => p.value,
        None => panic!("null pointer dereference"),  // explicit, not UB
    }
}
